src/Tools/Source/CompilerGeneratorTools/Source/VisualBasicSyntaxGenerator/Grammar/GrammarGenerator.vb

```diff
@@ -0,0 +1,295 @@
+                    nodeKindList.Select(Function(nk) HandleNodeKind(nk)).Distinct()).Parenthesize()
+            End If
+        End If
+
+        Throw New NotImplementedException()
+    End Function
+
+    Private Function GetNonRootParent(n As ParseNodeKind) As ParseNodeStructure
+        Dim current = n.NodeStructure
+        While Not current.ParentStructure.IsRoot
+            current = current.ParentStructure
+        End While
+
+        Return current
+    End Function
+
+    Private Function HandleNodeKind(nodeKind As ParseNodeKind) As Production
+        If Not String.IsNullOrEmpty(nodeKind.TokenText) Then
+            If nodeKind.TokenText = "\" Then
+                Return New Production("'\\'")
+            ElseIf nodeKind.TokenText = "'" Then
+                Return New Production("'\''")
+            Else
+                Return New Production("'" + nodeKind.TokenText + "'")
+            End If
+        End If
+
+        Return RuleReference(nodeKind.NodeStructure.Name)
+    End Function
+
+    Private Sub ProcessRule(name As String, ByRef result As String)
+        If name <> "VisualBasicSyntaxNode" AndAlso _seen.Add(name) Then
+            Dim sorted = _nameToProductions(name).Distinct().OrderBy(Function(v) v)
+            If sorted.Any() Then
+                result += Environment.NewLine + RuleReference(name).Text + Environment.NewLine + "  : " +
+                          String.Join(Environment.NewLine + "  | ", sorted) + Environment.NewLine + "  ;" + Environment.NewLine
+
+                ' Now proceed in depth-first fashion through the referenced rules to keep related rules
+                ' close by. Don't recurse into major-sections to help keep them separated in grammar file.
+                Dim references = sorted.SelectMany(Function(t) t.ReferencedRules)
+                For Each referencedRule In references.Where(Function(r) Not _majorRules.Concat(_lexicalRules).Contains(r))
+                    ProcessRule(referencedRule, result)
+                Next
+            End If
+        End If
+    End Sub
+
+    Private Function RuleReference(name As String) As Production
+        Dim trimmed = If(name.EndsWith("Syntax"), name.Substring(0, name.Length - "Syntax".Length), name)
+        Return New Production(
+            s_normalizationRegex.Replace(trimmed, "_").ToLower(),
+            ImmutableArray.Create(name))
+    End Function
+
+    Private Shared Function GetSyntaxKind(name As String) As SyntaxKind
+        Return GetMembers(Of SyntaxKind)().Where(Function(k) k.ToString() = name).SingleOrDefault()
+    End Function
+
+    Private Shared Function GetMembers(Of TEnum)() As IEnumerable(Of TEnum)
+        Return GetType(TEnum).GetFields(BindingFlags.Public Or BindingFlags.Static).
+            Select(Function(f) f.GetValue(Nothing)).OfType(Of TEnum)()
+    End Function
+
+    ' Converts a PascalCased name into snake_cased name.
+    Private Shared ReadOnly s_normalizationRegex As New Regex(
```

Comment:
this part down is a copy from C#.

Reply:
Do you want to reference the C# assembly to get `Production` factored?

Reply:
personally i would prefer *not* to do that.  it seems a little hinky.  and i don' thtink it's worth adding a shared lib just to have this small amount of code shared.  Are you ok with that?